src/sharedHooks/useUploadObservations.js

```diff
@@ -1,42 +1,52 @@
 // @flow
 
 import { RealmContext } from "providers/contexts";
-import { useCallback, useEffect, useState } from "react";
+import { useEffect, useState } from "react";
 import Observation from "realmModels/Observation";
 import useApiToken from "sharedHooks/useApiToken";
 
 const { useRealm } = RealmContext;
 
 const useUploadObservations = ( allObsToUpload: Array<Object> ): Object => {
-  const [cancelUpload, setCancelUpload] = useState( false );
+  const [uploadInProgress, setUploadInProgress] = useState( false );
+  const [shouldUpload, setShouldUpload] = useState( false );
   const [currentUploadIndex, setCurrentUploadIndex] = useState( 0 );
   const realm = useRealm( );
   const apiToken = useApiToken( );
 
-  const handleClosePress = useCallback( ( ) => {
-    setCancelUpload( true );
-  }, [] );
+  const cleanup = ( ) => {
+    setUploadInProgress( false );
+    setShouldUpload( false );
```

Comment:
I think this is a good question, and I'm not convinced a reducer makes a huge difference here. I spent a fair bit of time playing around with things like [why-did-you-render](https://github.com/welldone-software/why-did-you-render) in Seek development, and mostly came to the conclusion that extra re-renders are not costly in RN ([and don't always result in DOM updates](https://kentcdodds.com/blog/fix-the-slow-render-before-you-fix-the-re-render#unnecessary-re-renders)) and it's not often worth optimizing for that.

I think of reducers as being really useful for code cleanup for very lengthy or complex components. It's useful when there are too many things happening to keep track of with multiple setStates, or when there are a lot of different switch cases where vastly different UIs are shown.

Does that make sense? Do you still have a different opinion? I think this is the type of thing we can totally form a team opinion on, instead of relying on our individual understandings.

Reply:
I agree in a sense that a reducer is helpful for complex state. For me personally, seeing three consecutive setStates is borderline too much. In class components in this particular case I would have had something like an initial state object that I would have reverted to with one setState call. This was not an opinion to change this PR though, just wondering about potential side-effects of "partial" re-render: after one setState passed and the others not yet.